table/column.go

```diff
@@ -177,10 +177,84 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		if err1 != nil {
 			logutil.BgLogger().Warn("Datum ToString failed", zap.Stringer("Datum", val), zap.Error(err1))
 		}
-		err = sc.HandleTruncate(types.ErrTruncatedWrongVal.GenWithStackByArgs(col.FieldType.CompactStr(), str))
-	} else {
-		err = sc.HandleTruncate(err)
+		err = types.ErrTruncatedWrongVal.GenWithStackByArgs(col.FieldType.CompactStr(), str)
+	} else if (sc.InInsertStmt || sc.InUpdateStmt) && !casted.IsNull() &&
+		(val.Kind() != types.KindMysqlTime || !val.GetMysqlTime().IsZero()) &&
```

Comment:
what about extract below code into a function?

Reply:
May I ask whether this piece of code https://github.com/pingcap/tidb/blob/bec1cc01b506ea94b86e105e8126de1d208b6bfa/table/column.go#L183-L254 should be included with a function? At present, there is a lot of code to handle Datetime in this part, and putting it in `CastValue` will affect the main logic. 

Good idea, I can put it into a function such as `castDatetime()`.

Reply:
yes, the code that you added can be put into a function, so we can make CastValue function more readable.